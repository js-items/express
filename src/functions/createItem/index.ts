import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { CREATED } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import sendResponse from '../../utils/sendResponse';

const createItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeCreateItem
  );

  await transactionHandler({ req, res }, async () => {
    const { item } = await config.service.createItem({
      id: req.body.id,
      item: config.convertDocumentIntoItem({ document: req.body, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item, req, res }),
      status: CREATED,
    });
  });
};

export default createItem;
