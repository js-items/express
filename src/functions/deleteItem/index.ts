import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { NO_CONTENT } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const deleteItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeDeleteItem
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');
    
    await config.service.deleteItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
    });

    sendResponse({
      config,
      req,
      res,
      status: NO_CONTENT,
    });
  });
};

export default deleteItem;
