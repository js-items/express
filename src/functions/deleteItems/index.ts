import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { NO_CONTENT } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const deleteItems: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeDeleteItems
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');
    
    await config.service.deleteItems({
      filter: config.createFilter({ filter, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      status: NO_CONTENT,
    });
  });
};

export default deleteItems;
