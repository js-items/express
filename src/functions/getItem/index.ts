import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const getItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeGetItem
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const { item } = await config.service.getItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
    });
    
    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item, req, res }),
      status: OK,
    });
  });
};

export default getItem;
