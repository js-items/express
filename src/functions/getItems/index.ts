import { Item } from '@js-items/foundation';
import boolean from 'boolean';
import { Request, Response } from 'express';
import { OutgoingHttpHeaders } from 'http';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import _isNil from 'ramda/src/isNil';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import getNumberQueryParam from '../../utils/getNumberQueryParam';
import sendResponse from '../../utils/sendResponse';

const getItems: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeGetItems
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const sort = !_isNil(req.query.sort)
      ? getJsonQueryParam(req.query, 'sort')
      : config.defaultSort;
      
    const limit = getNumberQueryParam(
      req.query,
      'limit',
      config.defaultPaginationLimit
    );
    
    const createdFilter = config.createFilter({ filter, req, res });

    const { cursor, items } = await config.service.getItems({
      filter: createdFilter,
      pagination: {
        after: req.query.after,
        before: req.query.before,
        limit,
      },
      sort,
    });

    const { count: totalCount } = await config.service.countItems({
      filter: createdFilter,
    });

    const responseHeaders: OutgoingHttpHeaders = {
      [config.afterHeaderName]: cursor.after,
      [config.beforeHeaderName]: cursor.before,
      [config.hasBeforeHeaderName]: cursor.hasBefore.toString(),
      [config.hasAfterHeaderName]: cursor.hasAfter.toString(),
      [config.totalHeaderName]: totalCount,
    };

    const responseData = items.map(item =>
      config.convertItemIntoDocument({ item, req, res })
    );

    const nestedObject = {
      [config.paginationKey]: {
        [config.afterKey]: cursor.after,
        [config.beforeKey]: cursor.before,
        [config.hasBeforeKey]: cursor.hasBefore,
        [config.hasAfterKey]: cursor.hasAfter,
        [config.totalKey]: totalCount,
      },
      [config.dataKeyName]: responseData,
    };

    const enveloped = boolean(req.query[config.envelopeParamName]);

    const responseObject = enveloped ? nestedObject : responseData;
    const headers = enveloped ? {} : responseHeaders;

    sendResponse({ req, res, config, status: OK, headers, responseObject });
  });
};

export default getItems;
