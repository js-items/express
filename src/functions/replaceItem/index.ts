import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const replaceItem: RequestHandlerFactory =  <I extends Item>(config: FacadeConfig<I>) => async (
  req: Request,
  res: Response
) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(config.beforeReplaceItem);

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const result = await config.service.replaceItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
      item: config.convertDocumentIntoItem({ document: req.body, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item: result.item, req, res }),
      status: OK,
    });
  });
};

export default replaceItem;