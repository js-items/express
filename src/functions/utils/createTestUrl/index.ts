// tslint:disable:no-any
import _mapObjIndexed from 'ramda/src/mapObjIndexed';

export interface Options {
  readonly url: string;
  readonly params?: { [paramName: string]: any };
}

const createTestUrl = ({ params = {}, url }: Options) =>
  Object.keys(params).reduce((acc, next) => {
    if (acc.indexOf('?') === -1) {
      return `${acc}?${next}=${params[next]}`;
    }

    return `${acc}&${next}=${params[next]}`;
  }, url);

export default createTestUrl;
