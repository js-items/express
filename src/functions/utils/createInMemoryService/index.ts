// tslint:disable:no-object-literal-type-assertion
import { Item } from '@js-items/foundation';
import Facade from '@js-items/foundation/dist/Facade';
import {
  Options as CreateItemOptions,
  Result as CreateItemResult,
} from '@js-items/foundation/dist/functions/CreateItem';
import {
  Options as GetItemOptions,
  Result as GetItemResult,
} from '@js-items/foundation/dist/functions/GetItem';
import { Result as GetItemsResult } from '@js-items/foundation/dist/functions/GetItems';
import { Result as ReplaceItemResult } from '@js-items/foundation/dist/functions/ReplaceItem';
import { Result as UpdateItemResult } from '@js-items/foundation/dist/functions/UpdateItem';
import { start } from '@js-items/foundation/dist/interfaces/Cursor';
import FactoryConfig from './FactoryConfig';

const defaultPromise = () => Promise.resolve();
/* istanbul ignore next */
const createInMemoryService = <I extends Item>(
  factoryConfig: FactoryConfig<I>
): Facade<I> => ({
  countItems: () => Promise.resolve({ count: 1 }),
  createItem: ({ item }: CreateItemOptions<I>) =>
    Promise.resolve({ item } as CreateItemResult<I>),
  deleteItem: defaultPromise,
  deleteItems: defaultPromise,
  getItem: ({ id }: GetItemOptions<I>) =>
    Promise.resolve({ item: { id } } as GetItemResult<I>),
  getItems: () =>
    Promise.resolve({
      cursor: {
        after: start,
        before: start,
        hasAfter: false,
        hasBefore: false,
      },
      items: [{ id: '1' }, { id: '2' }],
    } as GetItemsResult<I>),
  replaceItem: () =>
    Promise.resolve({ item: { id: '2' } } as ReplaceItemResult<I>),
  updateItem: () =>
    Promise.resolve({ item: { id: '1' } } as UpdateItemResult<I>),
  ...factoryConfig,
});

export default createInMemoryService;
