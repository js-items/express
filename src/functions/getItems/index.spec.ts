// tslint:disable:no-object-literal-type-assertion
import * as sourceMapSupport from 'source-map-support';
sourceMapSupport.install();
import dotenv from 'dotenv';
import { BAD_REQUEST, OK } from 'http-status-codes';
dotenv.config();
import { Result as GetItemsResult } from '@js-items/foundation/dist/functions/GetItems';
import { TestItem } from '@js-items/foundation/dist/functions/utils/testItem';
import { start } from '@js-items/foundation/dist/interfaces/Cursor';
import { TEST_URL } from '../../constants';
import createInMemoryService from '../../utils/createInMemoryService';
import createTestUrl from '../../utils/createTestUrl';
import initTests from '../../utils/initTests';

jest.mock('uuid', () => ({
  v4: jest.fn(() => '1'),
}));

const getItems = jest.fn(() =>
  Promise.resolve({
    cursor: {
      after: start,
      before: start,
      hasAfter: false,
      hasBefore: false,
    },
    items: [{ id: '1' }, { id: '2' }],
  } as GetItemsResult<TestItem>)
);

describe('@getItems', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
  });

  const service = createInMemoryService({
    getItems,
  });

  const { request } = initTests<TestItem>({ service });

  const filter = { id: { $in: ['1', '2'] } };
  const sort = { id: 'desc' };
  const limit = '1';
  const after = 'eyJpZCI6ImlkMSJ9';

  const defaultParams = {
    after,
    filter: JSON.stringify(filter),
    limit,
    sort: JSON.stringify(sort),
  };

  it('gets items', async () => {
    const url = createTestUrl({ url: TEST_URL, params: defaultParams });

    const { status, body } = await request.get(url).send();

    expect(status).toBe(OK);
    expect(body).toMatchSnapshot();
    expect(getItems).toBeCalledWith({
      filter,
      pagination: {
        after,
        before: undefined,
        limit: 1,
      },
      sort,
    });
  });

  it('gets items when envelope enabled and count query param provided', async () => {
    const params = {
      ...defaultParams,
      count: true,
      envelope: true,
    };

    const url = createTestUrl({ url: TEST_URL, params });

    const { status, body } = await request.get(url).send();

    expect(status).toBe(OK);
    expect(body).toMatchSnapshot();
  });

  it('gets items when envelope enabled and pretty response is disabled', async () => {
    const params = {
      ...defaultParams,
      envelope: true,
      pretty: false,
    };

    const url = createTestUrl({ url: TEST_URL, params });

    const { status, body } = await request.get(url).send();

    expect(status).toBe(OK);
    expect(body).toMatchSnapshot();
  });

  it('throws JsonError when sort is invalid', async () => {
    const params = {
      sort: 'invalid_filter',
    };
    const url = createTestUrl({ url: TEST_URL, params });

    const { status, body } = await request.get(url).send();

    expect(status).toBe(BAD_REQUEST);
    expect(body).toMatchSnapshot();
  });

  it('throws NumberError when limit is invalid', async () => {
    const params = {
      limit: 'not_numeric',
    };
    const url = createTestUrl({ url: TEST_URL, params });

    const { status, body } = await request.get(url).send();

    expect(status).toBe(BAD_REQUEST);
    expect(body).toMatchSnapshot();
  });
  // tslint:disable-next-line:max-file-line-count
});
