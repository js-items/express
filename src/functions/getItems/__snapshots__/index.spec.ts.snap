// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@getItems gets items 1`] = `
Array [
  Object {
    "id": "1",
  },
  Object {
    "id": "2",
  },
]
`;

exports[`@getItems gets items when envelope enabled 1`] = `
Object {
  "headers": Object {
    "x-powered-by": "Express",
  },
  "response": Object {
    "data": Array [
      Object {
        "id": "1",
      },
      Object {
        "id": "2",
      },
    ],
    "pagination": Object {
      "has_after": false,
      "has_before": false,
      "total_count": 1,
    },
  },
  "status": 200,
}
`;

exports[`@getItems gets items when envelope enabled and pretty response is disabled 1`] = `
Object {
  "headers": Object {
    "x-powered-by": "Express",
  },
  "response": Object {
    "data": Array [
      Object {
        "id": "1",
      },
      Object {
        "id": "2",
      },
    ],
    "pagination": Object {
      "has_after": false,
      "has_before": false,
      "total_count": 1,
    },
  },
  "status": 200,
}
`;

exports[`@getItems throws JsonError when sort is invalid 1`] = `
Object {
  "data": "invalid_sort",
  "path": Array [
    "filter",
  ],
  "transactionId": "1",
}
`;

exports[`@getItems throws JsonError when sort is invalid 2`] = `
Object {
  "data": "invalid_sort",
  "path": Array [
    "sort",
  ],
  "transactionId": "1",
}
`;

exports[`@getItems throws NumberError when limit is invalid 1`] = `
Object {
  "data": "not_numeric",
  "path": Array [
    "limit",
  ],
  "transactionId": "1",
}
`;
